MagazineLayout/LayoutCore/ModelState.swift

```diff
@@ -711,6 +712,11 @@ final class ModelState {
 
   private var sectionMaxYsCache = [CGFloat?]()
 
+  private var headerLocationsForFlattenedIndices = [Int: ElementLocation]()
+  private var footerLocationsForFlattenedIndices = [Int: ElementLocation]()
+  private var backgroundLocationsForFlattenedIndices = [Int: ElementLocation]()
+  private var itemLocationsForFlattenedIndices = [Int: ElementLocation]()
```

Comment:
I don't love that we need these, so open to suggestions if I'm missing an obvious improvement. We need something to help us out in the binary search algorithm. To understand the problem, consider if we didn't have these, and instead, just had section indices with item counts for each:

Section 0, 10 items
Section 1, 1 item
Section 2, 1000 items
Section 3, 0 items

To do binary search, I need to start in the "middle" of these. Finding the middle, given just that data, seems non-trivial (certainly doable, but it seems expensive). By flattening the element locations into an array, the middle simply becomes the middle of the array.